pipicswd/doc/pipicswitch.1
.\" Automatically generated by Pod::Man 2.22 (Pod::Simple 3.13)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "PIPICSWITCH 1"
.TH PIPICSWITCH 1 "2014-02-15" "version 20140215" "Raspberry Pi"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
pipicswitch \-  control external switch board with i2c
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBpipicswitch\fR \fB\-a\fR i2c address [\fB\-s\fR 1|2:on|off|toggle] [\fB\-r\fR] [\fB\-t\fR] 
[\fB\-T\fR] [\fB\-h\fR] [\fB\-v\fR] [\fB\-V\fR]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The \fBpipicswitch\fR talks to a \s-1PIC\s0 processor on i2c bus connected to Raspberry 
Pi. If only i2c address is given the switch status is printed. The switch can
be operated with the \fB\-s\fR option.
.SH "OPTIONS"
.IX Header "OPTIONS"
\&\fB\-a\fR chip address on i2c bus
.PP
\&\fB\-s\fR channel number 1 or 2 followed by ':' and off/on/toggle
.PP
\&\fB\-r\fR reset \s-1PIC\s0 timer
.PP
\&\fB\-t\fR read \s-1PIC\s0 timer
.PP
\&\fB\-T\fR test i2c communication to the board
.PP
\&\fB\-h\fR display a short help text
.PP
\&\fB\-v\fR verbose
.PP
\&\fB\-V\fR print version
.SH "EXAMPLE"
.IX Header "EXAMPLE"
Print switch status
.PP
\&\fBpipicswitch\fR \fB\-a\fR 27 \fB\-v\fR
.PP
Switch channel 2 on
.PP
\&\fBpipicswitch\fR \fB\-a\fR 27 \fB\-s\fR 2:on
.SH "WARNING"
.IX Header "WARNING"
No check is done where the query data is written. Could make some hardware 
unusable.
.SH "AUTHORS"
.IX Header "AUTHORS"
Jaakko Koivuniemi
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIpipic\fR\|(1), \fIpipicfile\fR\|(1), \fIpipictest\fR\|(1), \fIi2cdetect\fR\|(8), \fIi2cset\fR\|(8), \fIi2cget\fR\|(8)


pipicswitch/pipicswitch.c
/**************************************************************************
 * 
 * Control external power switch using i2c bus on Raspberry Pi. The switch
 * has a PIC processor to interprete commands send by this program.  
 *       
 * Copyright (C) 2014 Jaakko Koivuniemi.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 ****************************************************************************
 *
 * Sat Feb 15 20:35:37 CET 2014
 * Edit: Sat Feb 15 23:47:13 CET 2014
 *
 * Jaakko Koivuniemi
 **/

#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <linux/i2c-dev.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/file.h>
#include <unistd.h>
#include <getopt.h>
#include <errno.h>
#include <time.h>

#define CHECK_BIT(var,pos) !!((var) & (1<<(pos)))

const int version=20140215; // program version

const char i2cdev[100]="/dev/i2c-1";
int  address=0x00;
const int  i2lockmax=10; // maximum number of times to try lock i2c port  

// write i2c command to PIC optionally followed by data, length is the number 
// of bytes and can be 0, 1, 2 or 4 
// return: 1=ok, -1=open failed, -2=lock failed, -3=bus access failed, 
// -4=i2c slave writing failed
int write_cmd(int cmd, int data, int length, int verb)
{
  int ok=0;
  int fd,rd;
  int rnxt=0;
  int cnt=0;
  unsigned char buf[10];

  if((cmd>=0)&&(cmd<=255))
  {
    if((fd=open(i2cdev, O_RDWR)) < 0) 
    {
      printf("Failed to open i2c port\n");
      return -1;
    }

    rd=flock(fd, LOCK_EX|LOCK_NB);

    cnt=i2lockmax;
    while((rd==1)&&(cnt>0)) // try again if port locking failed
    {
      sleep(1);
      rd=flock(fd, LOCK_EX|LOCK_NB);
      cnt--;
    }

    if(rd)
    {
      printf("Failed to lock i2c port\n");
      return -2;
    }

    if(ioctl(fd, I2C_SLAVE, address) < 0) 
    {
      printf("Unable to get bus access to talk to slave\n");
      return -3;
    }

    buf[0]=cmd;
    if(length==1)
    { 
      buf[1]=data;
      if(verb==1) printf("Send 0x%02x%02x\n",buf[0],buf[1]);
      if((write(fd, buf, 2)) != 2) 
      {
        printf("Error writing to i2c slave\n");
        return -4;
      }
   }
   else if(length==2)
   { 
      buf[1]=(int)(data/256);
      buf[2]=data%256;
      if(verb==1) printf("Send 0x%02x%02x%02x\n",buf[0],buf[1],buf[2]);
      if((write(fd, buf, 3)) != 3) 
      {
        printf("Error writing to i2c slave\n");
        return -4;
      }
   }
   else if(length==4)
   { 
      buf[1]=(int)(data/16777216);
      rnxt=data%16777216;
      buf[2]=(int)(rnxt/65536);
      rnxt=rnxt%65536;
      buf[3]=(int)(rnxt/256);
      buf[4]=rnxt%256;
      if(verb==1) printf("Send 0x%02x%02x%02x%02x%02x\n",buf[0],buf[1],buf[2],buf[3],buf[4]);
      if((write(fd, buf, 5)) != 5) 
      {
        printf("Error writing to i2c slave\n");
        return -4;
      }
   }
   else
   {
      if(verb==1) printf("Send 0x%02x\n",buf[0]);
      if((write(fd, buf, 1)) != 1) 
      {
        printf("Error writing to i2c slave\n");
        return -4;
      }
   }
   close(fd);
   ok=1;
  }
  
  return ok;
}

// read data with i2c from PIC, length is the number of bytes to read 
// return: -1=open failed, -2=lock failed, -3=bus access failed, 
// -4=i2c slave reading failed
int read_data(int length, int verb)
{
  int rdata=0;
  int fd,rd;
  int cnt=0;
  unsigned char buf[10];

  if((fd=open(i2cdev, O_RDWR)) < 0) 
  {
    printf("Failed to open i2c port\n");
    return -1;
  }

  rd=flock(fd, LOCK_EX|LOCK_NB);

  cnt=i2lockmax;
  while((rd==1)&&(cnt>0)) // try again if port locking failed
  {
    sleep(1);
    rd=flock(fd, LOCK_EX|LOCK_NB);
    cnt--;
  }

  if(rd)
  {
    printf("Failed to lock i2c port\n");
    return -2;
  }

  if(ioctl(fd, I2C_SLAVE, address) < 0) 
  {
    printf("Unable to get bus access to talk to slave\n");
    return -3;
  }

  if(length==1)
  {
     if(read(fd, buf,1)!=1) 
     {
       printf("Unable to read from slave\n");
       return -4;
     }
     else 
     {
       if(verb==1) printf("Receive 0x%02x\n",buf[0]);
       rdata=buf[0];
     }
  } 
  else if(length==2)
  {
     if(read(fd, buf,2)!=2) 
     {
       printf("Unable to read from slave\n");
       return -4;
     }
     else 
     {
       if(verb==1) printf("Receive 0x%02x%02x\n",buf[0],buf[1]);
       rdata=256*buf[0]+buf[1];
     }
  }
  else if(length==4)
  {
     if(read(fd, buf,4)!=4) 
     {
       printf("Unable to read from slave\n");
       return -4;
     }
     else 
     {
        if(verb==1) printf("Receive 0x%02x%02x%02x%02x\n",buf[0],buf[1],buf[2],buf[3]);
        rdata=16777216*buf[0]+65536*buf[1]+256*buf[2]+buf[3];
     }
  }

  close(fd);

  return rdata;
}

// send 4 test bytes to PiPIC and read them back 
int testi2c(int verb)
{
  int ok=-1;
  int testint=0;
  int testres=1;

  srand((unsigned int)time(NULL));

  testint=rand();
  while((testint==-1)||(testint==-2)||(testint==-3)||(testint==-4))
  {
    testint=rand();
  }

  ok=write_cmd(0x02,testint,4,verb);

  if(ok!=1)
  {
    printf("failed to write 4 test bytes\n"); 
  }
  else
  {
    testres=read_data(4,verb);

    if((testres==-1)||(testres==-2)||(testres==-3)||(testres==-4))
    {
      printf("failed to read 4 test bytes\n"); 
    }
    else
    {
      if(testint==testres) ok=1; else ok=0;
    }
  }

  return ok;
}

// read internal timer 
int read_timer(int verb)
{
  int timer=-1;

  if(write_cmd(0x51,0,0,verb)!=1) printf("failed to send read timer command\n");
  else timer=read_data(4,verb);

  return timer;
}

// reset timer
int resetimer(int verb)
{
  int ok=-1;

  ok=write_cmd(0x50,0,0,verb);

  return ok;
}

void printusage()
{
  printf("usage: pipicswitch -a address [-s 1|2:on|off|toggle] [-r] [-t] [-T] [-h] [-v] [-V]\n");
}

void printversion()
{
  printf("pipicswitch v. 20140215, Jaakko Koivuniemi\n");
}

int main(int argc, char **argv)
{  
  int verb=0; // 1=verbosed output
  int i2ctest=0; // test i2c data flow
  int restimer=0; // reset PIC timer
  int readtimer=0; // read PIC timer
  int swtch=0; // operate switch
  int channel=0; // switch channel
  char operation[200]=""; // switch operation off/on/toggle

  int timer=0;
  int gpio=0;
  int ok=0;

  int optch=0;
  while(optch!=-1)
    {
      optch=getopt(argc,argv,"a:s:rtThvV");
      if(optch=='a')
        {
          sscanf(optarg,"%X",&address);
        }
      if(optch=='s')
        {
          if(sscanf(optarg,"%d:%s",&channel,operation)!=EOF) swtch=1;
        }
      if(optch=='T')
        {
          i2ctest=1;          
        }
      if(optch=='r')
        {
          restimer=1;          
        }
      if(optch=='t')
        {
          readtimer=1;          
        }
      if(optch=='v')
        {
          verb=1;
        }
      if(optch=='h')
        {
          printusage();
          return 0;
        }
      if(optch=='V')
        {
          printversion();
          return 0;
        }
    }

  if((address<0x03)||(address>0x77))
    {
      printusage();
      return -1;
    }
  else if(swtch==1)
    {
      if(channel==1)
        {
          if(strncmp(operation,"off",3)==0)
          {
            ok=write_cmd(0x14,0x00,0,verb); 
            if(ok!=1) printf("command failed\n");
          }
          else if(strncmp(operation,"on",2)==0)
          {
            ok=write_cmd(0x24,0x00,0,verb); 
            if(ok!=1) printf("command failed\n");
          }
          else if(strncmp(operation,"toggle",6)==0)
          {
            ok=write_cmd(0x34,0x10,1,verb); 
            if(ok!=1) printf("command failed\n");
          }
        }
      else if(channel==2)
        {
          if(strncmp(operation,"off",3)==0)
          {
            ok=write_cmd(0x15,0x00,0,verb); 
            if(ok!=1) printf("command failed\n");
          }
          else if(strncmp(operation,"on",2)==0)
          {
            ok=write_cmd(0x25,0x00,0,verb); 
            if(ok!=1) printf("command failed\n");
          }
          else if(strncmp(operation,"toggle",6)==0)
          {
            ok=write_cmd(0x34,0x20,1,verb); 
            if(ok!=1) printf("command failed\n");
          }
        }
    }
  else if(i2ctest==1)
    {
      ok=testi2c(verb);
      printf("%d",ok);
    }
  else if(restimer==1)
    {
      ok|=resetimer(verb);
    }
  else if(readtimer==1)
    {
      timer=read_timer(verb);
      printf("%d\n",timer);
    }
  else
    {
      // read GPIO
      ok=write_cmd(0x01,0x05,1,verb); 
      if(ok==1)
        { 
           gpio=read_data(1,verb);
           if(CHECK_BIT(gpio,4)==1) printf("switch 1 on\n");
           else printf("switch 1 off\n");
           if(CHECK_BIT(gpio,5)==1) printf("switch 2 on\n");
           else printf("switch 2 off\n");
        }
    }

  return ok;
}
